### PR TITLE
Delay import of DandiIcephysMetadata

### PR DESCRIPTION
**Pull request by h-mayorquin**
Imports should be done at __init__.

This is better for performance.

### CODE REVIEW

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 1]
```diff
@@ -33,6 +25,16 @@ def get_source_schema(cls) -> dict:
         return source_schema
 
     def __init__(self, file_paths: list):
+
+        try:
+            from ndx_dandi_icephys import DandiIcephysMetadata
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:30]: While we're here, can we replace the antiquated try/except block with `if < importlib? >.find_spec`?

Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:30]: I am lazy to do it but is the right way.

Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/icephys/baseicephysinterface.py]: OK, changed to an importlib findspec based solution.

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 2]
```diff
@@ -33,6 +26,17 @@ def get_source_schema(cls) -> dict:
         return source_schema
 
     def __init__(self, file_paths: list):
+        # Check if the ndx_dandi_icephys module is available
+        dandi_icephys_spec = importlib.util.find_spec("ndx_dandi_icephys")
+        if dandi_icephys_spec is not None:
+            dandi_icephys_module = importlib.util.module_from_spec(dandi_icephys_spec)
+            dandi_icephys_spec.loader.exec_module(dandi_icephys_module)
+            self.DandiIcephysMetadata = getattr(dandi_icephys_module, "DandiIcephysMetadata")
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:34]:
```suggestion
            from ndx_dandi_icephys import DandiIcephysMetadata
```

I mean just to determine if the package is installed; classic import syntax is cleaner than the rest once you know it's safe to call

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 3]
```diff
@@ -33,6 +26,17 @@ def get_source_schema(cls) -> dict:
...
+            self.HAVE_NDX_DANDI_ICEPHYS = True
+        else:
+            self.DandiIcephysMetadata = None
+            self.HAVE_NDX_DANDI_ICEPHYS = False
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:38]: Curious what you would think about setting these as the default at the class level (next to `ExtractorModuleName`)?

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py (hunk 2 above)
Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:34]: Sure.

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py (hunk 3 above)
Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:38]: I don't like it really. They are determined at import times, how we could make attributes of them?

**Review by CodyCBakerPhD — APPROVED** [review]

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 4]
```diff
@@ -33,6 +26,15 @@ def get_source_schema(cls) -> dict:
         return source_schema
 
     def __init__(self, file_paths: list):
+        # Check if the ndx_dandi_icephys module is available
+        dandi_icephys_spec = importlib.util.find_spec("ndx_dandi_icephys")
+        if dandi_icephys_spec is not None:
+
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:32]:
```suggestion
        if dandi_icephys_spec is not None:
            from ndx_dandi_icephys import DandiIcephysMetadata
            
            self.DandiIcephysMetadata = DandiIcephysMetadata

```

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 5]
```diff
@@ -49,8 +51,8 @@ def __init__(self, file_paths: list):
 
     def get_metadata_schema(self) -> dict:
         metadata_schema = super().get_metadata_schema()
-        if DandiIcephysMetadata:
-            metadata_schema["properties"]["ndx-dandi-icephys"] = get_schema_from_hdmf_class(DandiIcephysMetadata)
+        if self.DandiIcephysMetadata:
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:54]:
```suggestion
        if self.DandiIcephysMetadata is not None:
```

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py [hunk 6]
```diff
@@ -109,11 +111,11 @@ def add_to_nwbfile(
             nwbfile = make_nwbfile_from_metadata(metadata)
 
         if (
-            HAVE_NDX_DANDI_ICEPHYS
+            self.HAVE_NDX_DANDI_ICEPHYS
```
Comment (CodyCBakerPhD) [inline src/neuroconv/datainterfaces/icephys/baseicephysinterface.py:114]:
```suggestion
            self.HAVE_NDX_DANDI_ICEPHYS is not None
```

File: src/neuroconv/datainterfaces/icephys/baseicephysinterface.py (hunk 4 above)
Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/icephys/baseicephysinterface.py]: Thanks for fixing this. Went to eat and then got into a meeting.

**Pull request merged**